File: common/src/main/java/generations/gg/generations/core/generationscore/world/item/EnchantableItem.java
package generations.gg.generations.core.generationscore.world.item;

import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.ItemStack;
import org.jetbrains.annotations.NotNull;

public abstract class EnchantableItem extends Item {
    public EnchantableItem(Properties arg) {
        super(arg);
    }

    public int neededEnchantmentLevel(Player player) {
        return 100;
    }

    public static ItemStack getEnchanted(Item item) {
        ItemStack stack = new ItemStack(item);
        stack.getOrCreateTag().putBoolean("enchanted", true);
        return stack;
    }

    public static ItemStack setEnchanted(ItemStack stack, boolean enchanted) {
        stack.getOrCreateTag().putBoolean("enchanted", enchanted);
        return stack;
    }

    public static boolean isEnchanted(ItemStack stack) {
        return stack != null && !stack.isEmpty() && stack.getOrCreateTag().getBoolean("enchanted");
    }

    public boolean isFoil(@NotNull ItemStack stack) {
        return isEnchanted(stack);
    }
}

File: src/common/main/kotlin/generations/gg/generations/core/generationscore/common/world/item/legends/EnchantableItem.kt
package generations.gg.generations.core.generationscore.common.world.item.legends

import generations.gg.generations.core.generationscore.common.world.item.ItemWithLangTooltipImpl
import generations.gg.generations.core.generationscore.common.world.item.components.GenerationsDataComponents
import net.minecraft.server.level.ServerPlayer
import net.minecraft.world.item.Item
import net.minecraft.world.item.ItemStack

abstract class EnchantableItem(arg: Properties) : ItemWithLangTooltipImpl(arg.component(GenerationsDataComponents.ENCHANTED.value(), false)) {
    open fun neededEnchantmentLevel(player: ServerPlayer): Int {
        return 100
    }

    override fun isFoil(stack: ItemStack): Boolean {
        return isEnchanted(stack)
    }

    companion object {
        fun getEnchanted(item: Item): ItemStack {
            val stack = ItemStack(item)
            stack.set(GenerationsDataComponents.ENCHANTED.value(), true)
            return stack
        }

        @JvmStatic
        fun setEnchanted(stack: ItemStack, enchanted: Boolean): ItemStack {
            stack.set(GenerationsDataComponents.ENCHANTED.value(), enchanted)
            return stack
        }

        fun setUsed(stack: ItemStack, used: Boolean): ItemStack {
            stack.set(GenerationsDataComponents.USED.value(), used)
            return stack
        }

        @JvmStatic
        fun isUsed(stack: ItemStack?): Boolean {
            return stack != null && !stack.isEmpty && stack.getOrDefault(GenerationsDataComponents.USED.value(), false)
        }

        @JvmStatic
        fun isEnchanted(stack: ItemStack?): Boolean {
            return stack != null && !stack.isEmpty && stack.getOrDefault(GenerationsDataComponents.ENCHANTED.value(), false)
        }
    }
}
